### PR TITLE
[5.0] Fix Unit Testing Facade Issues

### LINKED ISSUE #7367: Consecutive requests in unit tests have the same input
Hi,

I have set up a very simple controller in a fresh installation of L5. Here's the code:

```php
namespace App\Http\Controllers;

use Illuminate\Support\Facades\Request;

class DummyController extends Controller {
	public function index() {
		var_dump(Request::all());
	}
}
```

I also have a test which performs two consecutive requests to this controller. The first one provides some parameters through the query string, whereas the second one does not:

```php
class DummyTest extends TestCase {
	public function testDummy() {
		$response = $this->call('GET', 'dummy?foo=bar&baz=quux');
		$response = $this->call('GET', 'dummy');
	}
}
```

Here's the route:

```php
Route::resource('dummy', 'DummyController', ['only' => 'index']);
```

When I run the test, I'm expecting the second `var_dump()` to be empty, however this is the output that I get:

```
$ vendor/bin/phpunit --filter DummyTest
PHPUnit 4.5.0 by Sebastian Bergmann and contributors.

.array(2) {
  ["foo"]=>
  string(3) "bar"
  ["baz"]=>
  string(4) "quux"
}
array(2) {
  ["foo"]=>
  string(3) "bar"
  ["baz"]=>
  string(4) "quux"
}


Time: 73 ms, Memory: 10.50Mb
```

I also came across this issue in L4, but I managed to solve it by restarting the client between requests:

```php
$response = $this->call('GET', 'dummy?foo=bar&baz=quux');
$this->client->restart();
$response = $this->call('GET', 'dummy');
```

However this no longer works in L5 since `$client` does not exist anymore.

I have been tracking the request throughout the call stack and everything seems to be OK right until the controller is actually called, so to me it looks like it has something to do with the Request object itself. It's like it is caching previous input for some reason. Also, `$this->refreshApplication()` works but it also deletes any in-memory database that I might be using.

Is this behavior correct?

### PR DESCRIPTION
**Pull request by janhenkgerritsen**
This pull request is a fix for the problem described in #7367.

### CODE REVIEW

File: src/Illuminate/Foundation/Testing/ApplicationTrait.php [hunk 1]
```diff
@@ -47,7 +48,13 @@ public function call($method, $uri, $parameters = [], $cookies = [], $files = []
 	{
 		$request = Request::create($uri, $method, $parameters, $cookies, $files, $server, $content);
 
-		return $this->response = $this->app->make('Illuminate\Contracts\Http\Kernel')->handle($request);
+		$this->response = $this->app->make('Illuminate\Contracts\Http\Kernel')->handle($request);
+
+		// Clear the resolved request instance, otherwise subsequent invocations of call()
+		// in the same test will reuse the first request instance
+		Facade::clearResolvedInstance('request');
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Testing/ApplicationTrait.php]: We should probably flush all cached instances from facades shouldn't we?

**Comment by GrahamCampbell** [comment]
:+1:

**Comment by taylorotwell** [comment]
Would it be better to clear them before calling the kernel rather than after?

**Comment by GrahamCampbell** [comment]
Maybe. Not sure if that would be breaking for anyone?

**Comment by crynobone** [comment]
Doing it before Kernel should be fine, however we need to reset app state to not has been bootstrapped in order for all bootstrap class to be executed again.

**Comment by taylorotwell** [comment]
Didn't quite follow that comment @crynobone

**Comment by crynobone** [comment]
See https://github.com/laravel/framework/blob/5.0/src/Illuminate/Foundation/Http/Kernel.php#L142

After the first request if we flush resolved instance and tun kernel again, none of the bootstrap class will be loaded since app has been bootstrapped before.

**Comment by taylorotwell** [comment]
Ah, I'm not sure if we care about that. Probably best bootstrappers only once?

**Comment by crynobone** [comment]
Might be just overthinking this, haha.

Another concern is how would you define mocked facade between both request?

**Comment by taylorotwell** [comment]
Solved this at the HTTP handler level.

**Pull request closed without merging**